In [314]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as pl

In [315]:
print(pd.__version__)  # 2.1.0

2.3.3


# **Ejercicio 1**

In [316]:
df = pd.read_csv('clima_talca_enero2024.csv')

In [317]:
df.head(5) 

,fecha,temp_max,temp_min,humedad_pct,precipitacion_mm,condicion
0,2024-01-01,32.1,18.2,45.0,0.0,Soleado
1,2024-01-02,33.5,17.5,48.0,0.0,Soleado
2,2024-01-03,NaN,19.0,42.0,0.0,Parcial
3,2024-01-04,31.8,NaN,50.0,0.0,Soleado
4,2024-01-05,34.2,17.8,NaN,0.0,Soleado


In [318]:
df.tail(3)

,fecha,temp_max,temp_min,humedad_pct,precipitacion_mm,condicion
28,2024-01-29,31.4,17.2,49.0,0.0,Parcial
29,2024-01-30,30.7,16.9,51.0,0.0,Nublado
30,2024-01-31,29.9,17.4,53.0,0.0,Soleado


In [319]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31 entries, 0 to 30
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   fecha             31 non-null     object 
 1   temp_max          25 non-null     float64
 2   temp_min          25 non-null     float64
 3   humedad_pct       28 non-null     float64
 4   precipitacion_mm  28 non-null     float64
 5   condicion         29 non-null     object 
dtypes: float64(4), object(2)
memory usage: 1.6+ KB


In [320]:
df.describe()

,temp_max,temp_min,humedad_pct,precipitacion_mm
count,25.000000,25.00000,28.000000,28.000000
mean,32.236000,17.84000,49.107143,0.321429
std,1.936896,0.87892,4.433369,0.840792
min,28.800000,16.20000,42.000000,0.000000
25%,30.700000,17.20000,45.750000,0.000000
50%,32.100000,17.80000,49.000000,0.000000
75%,33.700000,18.50000,52.250000,0.000000
max,35.400000,19.30000,58.000000,3.200000


In [321]:
df.shape

(31, 6)

### P1.1 ¿Cuántas filas y columnas tiene el dataset?

#### *df.tail muestra que la ultima fila es la numero 30, por lo tanto contando la fila 0 serian 31, ademas df.info respalda las 31 entradas*
#### *df.info dice que en total son 6 columnas*

### P1.2 ¿Qué columnas contienen datos de tipo float? ¿Cuál es la temperatura máxima promedio según describe()? 

#### *temp_max, temp_min, humedad_pct y precipitacion_mm son tipo float, segun describe la temperatura maxima promedio es 32.236000*

### P1.3 ¿Qué información relevante entrega df.info() que NO entrega df.describe()? 

#### *es mas explicito, entrega class, RangeIndex(entradas), Data column(columas) de manera directa, dytypes(cuantos hay y que columnas son), Non-Null Count y memory usage*

# **Ejercicio 2**

In [322]:
nulos_columna = df.isnull().sum()
porcentaje_nulos = (df.isnull().sum() / len(df)) * 100

resumen_nulos = pd.DataFrame({
    "Nulos": nulos_columna,
    "Porcentaje (%)": porcentaje_nulos
})

print("nulos columna",nulos_columna)
print("porcentaje_nulos",porcentaje_nulos)
print("qué columnas tienen más del 15%",resumen_nulos[resumen_nulos["Porcentaje (%)"] > 15])

nulos columna fecha               0
temp_max            6
temp_min            6
humedad_pct         3
precipitacion_mm    3
condicion           2
dtype: int64
porcentaje_nulos fecha                0.000000
temp_max            19.354839
temp_min            19.354839
humedad_pct          9.677419
precipitacion_mm     9.677419
condicion            6.451613
dtype: float64
qué columnas tienen más del 15%           Nulos  Porcentaje (%)
temp_max      6       19.354839
temp_min      6       19.354839


In [323]:
df.isnull().sum().sum()

np.int64(20)

### P2.1 ¿Cuántos valores faltantes tiene el dataset en total?

#### *20 valores faltantes*

### P2.2 ¿Qué columnas superan el 15% de nulos? ¿Qué implicancia tiene esto para el análisis? 

#### *Las columnas temp_max y temp_min superan el 15% de nulos*
#### *son las variables principales del análisis climático si se eliminan se perderian muchos datos*

### P2.3 ¿Por qué es importante calcular el porcentaje y no sólo el conteo absoluto de nulos?

#### *el conteo puede ser engañoso, medir con porcentajes nos hace tener en cuenta la relacion de los datos con el tamaño del data set*

# **Ejercicio 3**

In [324]:
dflimpio = df.copy()

dflimpio["temp_max"] = dflimpio["temp_max"].fillna(dflimpio["temp_max"].mean())
dflimpio["temp_min"] = dflimpio["temp_min"].fillna(dflimpio["temp_min"].mean())

dflimpio["humedad_pct"] = dflimpio["humedad_pct"].fillna(dflimpio["humedad_pct"].median())
dflimpio["precipitacion_mm"] = dflimpio["precipitacion_mm"].fillna(0.0)

filas_antes = df.shape[0]
dflimpio = dflimpio.dropna(subset=["condicion"])

filas_despues = dflimpio.shape[0]
print("Filas eliminadas:", filas_antes - filas_despues)
assert dflimpio.isnull().sum().sum() == 0

Filas eliminadas: 2


In [325]:
df.dropna().shape

(16, 6)

### P3.1 ¿Por qué se usó fillna(mean()) para temp_max y fillna(median()) para humedad_pct?

#### *por que la media representa bien a la temperatura al ser mas estable, en cambio la mediana es mejor para valores que pueden ser extremos como en la humedad*

### P3.2 ¿Cuántas filas se eliminaron al aplicar dropna(subset=['condicion'])? ¿Es aceptable esta pérdida? 

#### *2 filas, es aceptable la perdida ya que es baja respecto al total del data set*

### P3.3 ¿Qué sucede si se aplica df.dropna() sin parámetros? ¿Cuántas filas quedarían? Pruébalo.

#### *elimina las filas que tienen un dato nulo, quedando 16 filas*

# **Ejercicio 4**

In [326]:
print("\n la fecha, temp_max y condicion:")
print(dflimpio[["fecha", "temp_max", "condicion"]])


 la fecha, temp_max y condicion:
         fecha  temp_max condicion
0   2024-01-01    32.100   Soleado
1   2024-01-02    33.500   Soleado
2   2024-01-03    32.236   Parcial
3   2024-01-04    31.800   Soleado
4   2024-01-05    34.200   Soleado
5   2024-01-06    35.000  Caluroso
6   2024-01-07    32.236  Caluroso
7   2024-01-08    33.100   Parcial
8   2024-01-09    30.500   Nublado
9   2024-01-10    29.800  Lluvioso
10  2024-01-11    31.200  Lluvioso
11  2024-01-12    32.236   Parcial
12  2024-01-13    33.700   Soleado
13  2024-01-14    34.900   Soleado
14  2024-01-15    35.400  Caluroso
15  2024-01-16    32.236  Caluroso
17  2024-01-18    31.000   Parcial
18  2024-01-19    30.200   Nublado
19  2024-01-20    29.500   Nublado
20  2024-01-21    28.800  Lluvioso
21  2024-01-22    32.236  Lluvioso
22  2024-01-23    31.500   Parcial
23  2024-01-24    33.200   Soleado
24  2024-01-25    34.000   Soleado
25  2024-01-26    35.100  Caluroso
26  2024-01-27    32.236  Caluroso
28  2024-01-29    31.

In [327]:
print("\nPrimeras 5 filas:")
print(dflimpio.iloc[:5])


Primeras 5 filas:
        fecha  temp_max  temp_min  humedad_pct  precipitacion_mm condicion
0  2024-01-01    32.100     18.20         45.0               0.0   Soleado
1  2024-01-02    33.500     17.50         48.0               0.0   Soleado
2  2024-01-03    32.236     19.00         42.0               0.0   Parcial
3  2024-01-04    31.800     17.84         50.0               0.0   Soleado
4  2024-01-05    34.200     17.80         49.0               0.0   Soleado


In [328]:
print("\nFila con indice 14:")
print(dflimpio.loc[14])


Fila con indice 14:
fecha               2024-01-15
temp_max                  35.4
temp_min                  19.1
humedad_pct               43.0
precipitacion_mm           0.0
condicion             Caluroso
Name: 14, dtype: object


In [329]:
print("\nDias con la temp_max >= 34:")
print(dflimpio[dflimpio["temp_max"] >= 34])


Dias con la temp_max >= 34:
         fecha  temp_max  temp_min  humedad_pct  precipitacion_mm condicion
4   2024-01-05      34.2      17.8         49.0               0.0   Soleado
5   2024-01-06      35.0      18.5         44.0               0.0  Caluroso
13  2024-01-14      34.9      18.7         49.0               0.0   Soleado
14  2024-01-15      35.4      19.1         43.0               0.0  Caluroso
24  2024-01-25      34.0      17.6         47.0               0.0   Soleado
25  2024-01-26      35.1      18.9         44.0               0.0  Caluroso


In [330]:
print("\nDías con lluviosos:")
print(dflimpio[dflimpio["condicion"] == "Lluvioso"])


Días con lluviosos:
         fecha  temp_max  temp_min  humedad_pct  precipitacion_mm condicion
9   2024-01-10    29.800      17.1         53.0               1.8  Lluvioso
10  2024-01-11    31.200      18.3         49.0               0.0  Lluvioso
20  2024-01-21    28.800      16.2         58.0               0.0  Lluvioso
21  2024-01-22    32.236      17.0         55.0               3.2  Lluvioso


In [331]:
print("\nDuas donde el temp_max > 30 y la humedad_pct < 50:")
print(dflimpio[(dflimpio["temp_max"] > 30) & (dflimpio["humedad_pct"] < 50)])


Duas donde el temp_max > 30 y la humedad_pct < 50:
         fecha  temp_max  temp_min  humedad_pct  precipitacion_mm condicion
0   2024-01-01    32.100     18.20         45.0               0.0   Soleado
1   2024-01-02    33.500     17.50         48.0               0.0   Soleado
2   2024-01-03    32.236     19.00         42.0               0.0   Parcial
4   2024-01-05    34.200     17.80         49.0               0.0   Soleado
5   2024-01-06    35.000     18.50         44.0               0.0  Caluroso
6   2024-01-07    32.236     19.20         47.0               0.0  Caluroso
10  2024-01-11    31.200     18.30         49.0               0.0  Lluvioso
11  2024-01-12    32.236     17.90         46.0               0.0   Parcial
12  2024-01-13    33.700     17.84         48.0               0.0   Soleado
13  2024-01-14    34.900     18.70         49.0               0.0   Soleado
14  2024-01-15    35.400     19.10         43.0               0.0  Caluroso
15  2024-01-16    32.236     18.00  

In [332]:
dflimpio["amplitud_termica"] = dflimpio["temp_max"] - dflimpio["temp_min"]

print("\nDataFrame con amplitud_termica:")
print(dflimpio[["fecha", "temp_max", "temp_min", "amplitud_termica"]])


DataFrame con amplitud_termica:
         fecha  temp_max  temp_min  amplitud_termica
0   2024-01-01    32.100     18.20            13.900
1   2024-01-02    33.500     17.50            16.000
2   2024-01-03    32.236     19.00            13.236
3   2024-01-04    31.800     17.84            13.960
4   2024-01-05    34.200     17.80            16.400
5   2024-01-06    35.000     18.50            16.500
6   2024-01-07    32.236     19.20            13.036
7   2024-01-08    33.100     17.84            15.260
8   2024-01-09    30.500     16.50            14.000
9   2024-01-10    29.800     17.10            12.700
10  2024-01-11    31.200     18.30            12.900
11  2024-01-12    32.236     17.90            14.336
12  2024-01-13    33.700     17.84            15.860
13  2024-01-14    34.900     18.70            16.200
14  2024-01-15    35.400     19.10            16.300
15  2024-01-16    32.236     18.00            14.236
17  2024-01-18    31.000     17.30            13.700
18  2024-01-1

In [333]:
filtro = dflimpio[dflimpio["temp_max"] >= 34]
filtro.shape[0]

6

In [334]:
filtro["condicion"].value_counts()

condicion
Soleado     3
Caluroso    3
Name: count, dtype: int64

In [335]:
dflimpio["amplitud_termica"] = dflimpio["temp_max"] - dflimpio["temp_min"]
dflimpio["amplitud_termica"].mean()
dflimpio["amplitud_termica"].max()
dflimpio.loc[dflimpio["amplitud_termica"].idxmax()]

fecha               2024-01-06
temp_max                  35.0
temp_min                  18.5
humedad_pct               44.0
precipitacion_mm           0.0
condicion             Caluroso
amplitud_termica          16.5
Name: 5, dtype: object

In [336]:
dflimpio["amplitud_termica"].mean()

np.float64(14.363999999999997)

In [337]:
(dflimpio["temp_max"] > 30) & (dflimpio["humedad_pct"] < 50)

0      True
1      True
2      True
3     False
4      True
5      True
6      True
7     False
8     False
9     False
10     True
11     True
12     True
13     True
14     True
15     True
17    False
18    False
19    False
20    False
21    False
22    False
23     True
24     True
25     True
26     True
28     True
29    False
30    False
dtype: bool

### P4.1 ¿Cuántos días tuvo una temperatura máxima igual o superior a 34°C? ¿Qué condición climática predominó esos días? 

#### *6 dias, hubieron 3 dias soleados y 3 calurosos*

### P4.2 ¿Cuál es la amplitud térmica promedio del mes? ¿Qué día tuvo la mayor amplitud? 

#### *la amplitud termica promedio fue de 14.36°C, el 2024-01-06 tuvo 16.5° de amplitud termica*

### P4.3 ¿Por qué se utilizan paréntesis adicionales en el filtro compuesto? ¿Qué ocurre si se omiten? 

#### *para agrupar bien las condiciones logicas, si se omiten cambiaria el orden y romperia la logica*

# **Ejercicio 5**

In [338]:
dia_mas_caluroso = dflimpio.loc[dflimpio["temp_max"].idxmax()]
print("El dia mas caluroso:")
print(dia_mas_caluroso)

El dia mas caluroso:
fecha               2024-01-15
temp_max                  35.4
temp_min                  19.1
humedad_pct               43.0
precipitacion_mm           0.0
condicion             Caluroso
amplitud_termica          16.3
Name: 14, dtype: object


In [339]:
dia_mas_lluvioso = dflimpio.loc[dflimpio["precipitacion_mm"].idxmax()]
print("\nDia con la mayor precipitación:")
print(dia_mas_lluvioso)


Dia con la mayor precipitación:
fecha               2024-01-22
temp_max                32.236
temp_min                  17.0
humedad_pct               55.0
precipitacion_mm           3.2
condicion             Lluvioso
amplitud_termica        15.236
Name: 21, dtype: object


In [340]:
conteo_condiciones = dflimpio["condicion"].value_counts()
print("\nCantidad de dias por las condiciones climaticas:")
print(conteo_condiciones)


Cantidad de dias por las condiciones climaticas:
condicion
Soleado     9
Parcial     6
Caluroso    6
Nublado     4
Lluvioso    4
Name: count, dtype: int64


In [341]:
temp_max_prom = dflimpio["temp_max"].mean()
temp_min_prom = dflimpio["temp_min"].mean()

print("\nTemperatura maxima del promedio:", temp_max_prom)
print("Tempratura minima del promedio:", temp_min_prom)


Temperatura maxima del promedio: 32.204
Tempratura minima del promedio: 17.839999999999996


In [342]:
duplicados = dflimpio.duplicated().sum()
print("\nCantidad de filas duplicadas:", duplicados)


Cantidad de filas duplicadas: 0


In [343]:
print("\nResumen de las estadistica:")
print(dflimpio.describe())


Resumen de las estadistica:
        temp_max   temp_min  humedad_pct  precipitacion_mm  amplitud_termica
count  29.000000  29.000000    29.000000         29.000000         29.000000
mean   32.204000  17.840000    49.172414          0.310345         14.364000
std     1.788428   0.813722     4.310030          0.827796          1.404882
min    28.800000  16.200000    42.000000          0.000000         12.000000
25%    31.000000  17.300000    46.000000          0.000000         13.236000
50%    32.236000  17.840000    49.000000          0.000000         14.000000
75%    33.500000  18.300000    52.000000          0.000000         15.860000
max    35.400000  19.300000    58.000000          3.200000         16.500000


In [344]:
dflimpio["condicion"].value_counts()

condicion
Soleado     9
Parcial     6
Caluroso    6
Nublado     4
Lluvioso    4
Name: count, dtype: int64

In [345]:
dflimpio["temp_min"].mean()

np.float64(17.839999999999996)

### P5.1 ¿Qué condición climática dominó en enero 2024 en Talca? ¿Es consistente con el clima real de la zona en verano? 

#### *en talca domino el clima soleado en enero del 2024, es consistente ya que el clima de talca es seco y caliente en verano*

### P5.2 ¿La temperatura mínima promedio es consistente con las temperaturas nocturnas de Talca en enero? Justifica. 

#### *la temperatura minima promedio de 17.839°C es consistente por que en los veranos de talca sigue haciendo calor en la noche*

### P5.3 Si necesitaras compartir este análisis con alguien sin conocimientos de Python, ¿cómo presentarías los resultados? ¿Qué información priorizarías? 

#### *Mostraria los datos en un ppt con graficos simples (como de barras)y un resumen facil de entender donde entrarian los datos mas interesantes como el promedio, maximos, minimos, frecuencias climaticas y evitaria usar lenguaje muy tecnico*